autoprecompiles/src/memory_optimizer.rs

```diff
@@ -230,35 +218,55 @@ impl<T: FieldElement, V: Hash + Eq + Clone + Ord + Display> MemoryAddressCompara
 
         let memory_addresses = addresses
             .map(|addr| {
-                let addr = addr.transform_var_type(&mut |v| v.into());
-                (
-                    addr.clone(),
-                    find_equivalent_expressions(&addr, &constraint_system),
-                )
+                // Represent an address as a list of expressions, one for each limb.
+                let addr = Self::boolean_extracted_address(addr);
+                let equivalent_expressions = addr
+                    .iter()
```

Comment:
Might be nitpicky, but we require Address to implement Eq and we actually use that above for `memory_contents`. Here we implicitly assume that `Address::into_iter()` returns an iterator whose Eq is equivalent to that of Address.

Reply:
Or we could derive Eq and Hash from the iterator. Not sure if that is actually less efficient.

Reply:
Hmm, I see, what about cf9217d6c892f4a95f4e31c95f893f86ce384c7a?

I'm not sure how I would enforce that the lengths are equal. I don't think it's possible with an associated constant and a const generic would leak all the way to the `Adapter`, which seems blown out of proportion. But lmk if you have an idea :)